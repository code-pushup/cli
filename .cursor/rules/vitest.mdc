---
description: 
globs: 
alwaysApply: false
---

In this document you can find information about the CLI test suite and guidelines for writing tests for the CLI.

# Test suites
On this project, we distinguish three types of functional tests by scope - unit tests, integration tests and E2E tests. All tests are written using [Vitest](mdc:https:/vitest.dev)

## Unit tests
- Their scope refers to an isolated unit of code.
- Aim to only test the internal logic and use mocks where needed in order to narrow down the testing scope (file system and other).
- These tests are located **next to their production code counterpart** and have the `.unit.test.ts` suffix.
- One can run unit tests for a given package via `npx nx run <package>:unit-test`
- One can run all unit tests via `npx nx run-many -t unit-test`

## Integration tests
- Their scope refers to a bigger unit of code in isolation or an integration with a 3rd-party tool.
- Aim to test the output of the unit as a whole. When communicating with a 3rd-party tool, only tests its usage in CodePushup.
- These tests are also located **next to their production code counterpart** and have the `.integration.test.ts` suffix.
- Run integration tests for a given package via `npx nx run <package>:integration-test`
- Run all integration tests via `npx nx run-many -t integration-test`

# Automation

### run afftected tests

The test suites can be triggered with:
- `nx run <package>:<unit-test|integration-test>`
- `nx <unit-test|integration-test> <package>`
- `nx run-many -t <unit-test|integration-test>`

The test tests can be configured with:
- `nx <unit-test|integration-test> <package> -t <test-name-pattern>` - fine-grained test selection
- `nx <unit-test|integration-test> <package> --coverage.enabled` - coverage report

# Testing utilities
In order to keep the maintenance cost lower, there is a folder dedicated to testing utilities called `testing`. Any fixtures, helper functions, constants or mocks usable by multiple tests should be added to the [test-utils](mdc:https:/github.com/code-pushup/cli/tree/main/testing/test-utils) subfolder.

The folder structure should be `fixtures` for fixed data or examples and `utils` for utility functions.

The exception to this may be a utility which is applicable only to the current package. In this case, each package shall use the `mocks` folder to mirror the hierarchy of the `testing-utils` - fixtures go into `fixtures` subfolder and functions into `utils` subfolder. TypeScript is set up to include this folder only for its test configuration.

## Mocking
Each test suite has a predefined set of setup files which mock appropriate parts of the logic. These mock setups may be found in the [setup](mdc:https:/github.com/code-pushup/cli/tree/main/testing/test-setup) folder of the `testing-utils` library.

One may apply them inside `vitest.(unit|integration|e2e).config.ts` in the following way:

```typescript
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: [
      // any setup files go here, e.g. for mocking the console
      '../../testing-utils/src/lib/setup/console.mock.ts',
    ],
  },
});
```

> [!TIP]
> There is an additional file [reset.mocks.ts](mdc:https:/github.com/code-pushup/cli/blob/main/testing-utils/src/lib/setup/reset.mocks.ts) which may be included in the `setupFiles` to automatically reset your mocks.

The main setup is currently divided between unit and integration tests in the following way:
- Unit tests always have the console, file system and current working directory (points to the in-memory file system) mocked.
- Integration tests always have the console mocked.

## Test outputs
In order to keep the repository clean, test outputs should target ignored folders, such as `node_modules` or `tmp`.

The `tmp` folder flow should be as follows:
- Any test that needs outputs should create its own subfolder under `tmp`, e.g. in `beforeAll` of a given test suite.
- After the test suite is run, this subfolder should be removed in `afterAll`.

> [!IMPORTANT]
> A test suite should not expect the `tmp` folder to be created. Therefore, the subfolder should be created via [mkdir](mdc:https:/nodejs.org/api/fs.html#fspromisesmkdirpath-options) with `{ recursive: true }`.
> Likewise, a test suite should not remove the `tmp` folder itself, only the subfolder needed for the given test suite.

> [!TIP]
> `setupTestFolder` and `teardownTestFolder` utility functions from [test-folder-setup](mdc:https:/github.com/code-pushup/cli/blob/main/testing/test-setup/src/lib/test-folder.setup.ts) may be used.

# Test quality

## Good practices

### Motivation
I found it best to imagine **I am writing tests for a colleague who hasn't implemented that feature**. They are most likely reviewing my code, debugging a failing test or checking the test coverage. So they want to either find out **what behaviour** is being tested (and **how**) or where the tested **values** come from. And the less time they spend on trying to understand the tests, the better.

### Principles for Writing Readable and Maintainable Tests

> [!IMPORTANT]
> Test code is read and maintained far more than it's written. Optimise for **DAMP** (Descriptive and Meaningful Phrases).

The primary goal of test code is **readability**. Other developers (not just you) will read and maintain these tests. Here are common pitfalls and their improved alternatives:

---

#### âœ… Test Descriptions

Test descriptions should **clearly communicate the use case** being tested. Avoid vague or abstract phrases â€” they donâ€™t help readers understand intent.

---

**ðŸ”´ BAD: Too vague**

//```ts
it('should return correct data');
//```

**âœ… GOOD: Descriptive and specific**

//```ts
it('should return the user object for a given user ID');
//```

---

**ðŸ”´ BAD: Abstract and unclear**

//```ts
it('when X is set to true, Y is returned');
//```

**âœ… GOOD: Explains the actual condition and outcome**

//```ts
it('should return the user object containing the passed filter property when doFilter is true');
//```

> A good test name should make the use case obvious without needing to read the implementation.

---

#### âœ… Keep describe blocks flat. Use them for Scoped Context only.

Use `describe` **only** to group tests that share setup or context. Donâ€™t nest for no reason.

---

**ðŸ”´ BAD: Unnecessary nesting**

//```ts
describe('UserService', () => {
  describe('getUser', () => {
    it('returns user', () => { /* ... */ });
  });
});
//```

**âœ… GOOD: Flat when no shared setup**

//```ts
describe('UserService.getUser', () => {
  it('returns user', () => { /* ... */ });
});
//```

**âœ… GOOD: Nest only with shared context**

//```ts
describe('UserService', () => {
  beforeEach(() => initDb());
  describe('getUser', () => {
    it('returns user', () => { /* ... */ });
  });
});
//```


#### âœ… Test Scope and Relevance

A test should focus on a **single, clearly defined use case**. Avoid mixing concerns or testing implementation details. Letâ€™s look at concise and intuitive examples:

---

**ðŸ”´ BAD: Mixing multiple responsibilities in one test**

//```ts
it('should retrieve the user and check details', () => {
  const user = getUserById('123');

  expect(user.name).toBe('Alice');           // retrieval
  expect(user.role).toBe('admin');           // role logic
  expect(user.permissions).toContain('edit'); // permission logic
});
//```

**âœ… GOOD: One responsibility per test**

//```ts
it('should return user by ID', () => {
  const user = getUserById('123');
  expect(user.name).toBe('Alice');
});

it('should assign admin role to user', () => {
  const user = getUserById('123');
  expect(user.role).toBe('admin');
});

it('should grant edit permission to admins', () => {
  const user = getUserById('123');
  expect(user.permissions).toContain('edit');
});
//```

---

**ðŸ”´ BAD: Testing implementation (how), not behavior**

//```ts
it('should charge the payment API', () => {
  const spy = vi.spyOn(paymentApi, 'charge'); // too focused on internals
  processPayment(order);

  expect(spy).toHaveBeenCalledWith(order.total);
});
//```

**âœ… GOOD: Test what matters to the user (the result)**

//```ts
it('should return success when payment is processed', () => {
  const result = processPayment(order);
  expect(result.status).toBe('success');
});
//```

> Keep tests laser-focused. Test *what* the system does, not *how* it does it.

---

#### Test data 

TODO: inline data, minimal data, no setup var data.

#### check data structures

TODO strict equal et  maybe already here

#### âœ… Choosing Between `it.each` and Separate `it` Blocks

Use `it.each` for simple, uniform cases; for complex objects write separate `it` blocks.

---

**ðŸ”´ BAD: Complex objects in `it.each`**

//```ts
it.each([
  [{ name: 'Alice', role: 'admin', perms: ['read','write'] }],
  [{ name: 'Bob',   role: 'guest', perms: ['read'] }],
])('createUser(%o)', async (user) => {
  const res = await createUser(user);
  expect(res).toMatchObject({ ...user, id: expect.any(String) });
});
//```

**âœ… GOOD: Separate `it` blocks for complex cases**

//```ts
it('creates admin with read/write perms', async () => {
  const admin = { name: 'Alice', role: 'admin', perms: ['read','write'] };
  await expect(createUser(admin)).resolves.toMatchObject({ ...admin, id: expect.any(String) });
});

it('creates guest with read-only perms', async () => {
  const guest = { name: 'Bob', role: 'guest', perms: ['read'] };
  await expect(createUser(guest)).resolves.toMatchObject({ ...guest, id: expect.any(String) });
});
//```

---

**ðŸ”´ BAD: Separate `it` blocks for regex**

//```ts
it('accepts valid com email', () => {
  expect(validateEmail('alice@example.com')).toBe(true);
});
it('accepts valid io email', () => {
  expect(validateEmail('alice@example.io')).toBe(false);
});
//```

**âœ… GOOD: `it.each` for regex cases**

//```ts
it.each([
  ['alice@example.com'],
  ['alice@example.io'],
])('validateEmail(%s) â†’ true', (input) => {
  expect(validateEmail(input)).toBe(true);
});
//```


#### âœ… Test Flow

Tests should be **linear** and **self-contained** â€” easy to read from top to bottom with no detours or mental overhead.

---

**ðŸ”´ BAD: Uses conditionals â€” adds unnecessary logic**

//```ts
it('should calculate discount', () => {
  const isPremium = true;
  const discount = isPremium ? 20 : 10;

  expect(applyDiscount(100, discount)).toBe(80);
});
//```

**âœ… GOOD: Clear and direct**

//```ts
it('should apply 20% discount for premium users', () => {
  expect(applyDiscount(100, 20)).toBe(80);
});
//```

---

**ðŸ”´ BAD: Depends on external helper â€” hides test context**

//```ts
it('should send welcome email', () => {
  const user = createTestUser(); // unclear details
  expectEmailSentTo(user.email);
});
//```

**âœ… GOOD: All data is in the test itself**

//```ts
it('should send welcome email', () => {
  const user = { email: 'alice@example.com' };
  expectEmailSentTo('alice@example.com');
});
//```

> Avoid logic and hidden setup. A test should **tell a story at a glance**.

---

#### âœ… Value Usage

Tests should use **realistic, descriptive, and direct values**. Avoid vague names and overly abstract helpers.

---

**ðŸ”´ BAD: Vague names, unclear intent**

//```ts
it('should return correct result', () => {
  const temp = getData(mockData);
  expect(result1).toEqual(temp);
});
//```

**âœ… GOOD: Clear and meaningful**

//```ts
it('should return active user', () => {
  const input = { id: '123', status: 'active' };
  expect(getUser(input).status).toBe('active');
});
//```

---

**ðŸ”´ BAD: Helper hides test logic**

//```ts
it('should reject invalid login', () => {
  const user = createInvalidUser();
  expect(authenticate(user)).toBe(false);
});
//```

**âœ… GOOD: Explicit and readable**

//```ts
it('should reject wrong password', () => {
  const user = { username: 'alice', password: 'wrongpass' };
  expect(authenticate(user)).toBe(false);
});
//```

> Prefer direct values over indirection. Clarity beats cleverness in tests.

---

#### âœ… Naming

Names should **immediately convey purpose**. Avoid generic terms that require context to understand.

---

**ðŸ”´ BAD**

//```ts
const data = getData();
const result = mock(data);
//```

**âœ… GOOD**

//```ts
const user = createUserWithEmail('alice@example.com');
const token = generateAdminToken(user);
//```

---

#### âœ… DRY in Balance with DAMP

---

**ðŸ”´ BAD: Repeated tests**

//```ts
it('adds 1+1', () => expect(add(1,1)).toBe(2));
it('adds 2+2', () => expect(add(2,2)).toBe(4));
//```

**âœ… GOOD: Dynamic tests with `it.each`**

//```ts
it.each([
  [1, 1, 2],
  [2, 2, 4],
])('add(%i, %i) â†’ %i', (a, b, expected) => {
  expect(add(a, b)).toBe(expected);
});
//```

---

**ðŸ”´ BAD: Duplicating setup in every file**

//```ts
import { initDb } from './db';
beforeEach(() => initDb());
//```

**âœ… GOOD: Centralize in config or utils**

//```ts
// vitest.config.ts
export default {
  setupFiles: ['./testing-utils/initDb.ts'],
};

---

**ðŸ”´ BAD: Same behavior tested in two places**

//```ts
// auth/user.spec.ts
it('validates email format', () => {
  expect(validateEmail('alice@example.com')).toBe(true);
});

// utils/email.spec.ts
it('validates email format', () => {
  expect(validateEmail('alice@example.com')).toBe(true);
});
//```

**âœ… GOOD: One focused test in a single spec**

//```ts
// tests/validateEmail.spec.ts
it('returns true for valid email addresses', () => {
  expect(validateEmail('alice@example.com')).toBe(true);
});
//```

---

By writing tests this way, you ensure they're easy to understand, maintain, and extend.


## ESLint rules

[![CI](mdc:https:/github.com/code-pushup/cli/actions/workflows/ci.yml/badge.svg?branch=main)](https://github.com/code-pushup/cli/actions/workflows/ci.yml?query=branch%3Amain)

Vitest is a powerful tool. Its methods should be fully utilised to serve the tests. In order to ensure the quality of our test code and promote Vitest's good practices, we have a set of ESLint rules in place for our test files. We use our own set of ESLint rules which can be found and reused from [@code-pushup/eslint-config](mdc:https:/github.com/code-pushup/eslint-config). The test rules can be found in [@code-pushup/eslint-config/vitest](mdc:https:/github.com/code-pushup/eslint-config/blob/main/docs/vitest.md).

Run ESLint on a specific package via `npx nx run <package>:lint`.

# Ai Prompts

fix tests after valid code changes 

```text
first nvm use 23 # A fix for the current setup. IT IS VERY IMPORTANT to have a working setup before using AI. Do not fix a setup in a prompt.

then
nx integration-test cpu-prof-e2e

for each test suggest a fix i can apporve.
Do not edit files without my approval.

Dont add commants. Never. 

I edited the code. so the tests will fail. The code is correct. Adjust the tests..
```

# TODO

nx integration-test cpu-prof --testFile src/lib/merge-cpuprofile-files.integration.test.ts -u

