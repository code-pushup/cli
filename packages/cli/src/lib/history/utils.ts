import { HistoryOptions } from '@code-pushup/core';
import { getHashFromTag, isSemver } from '@code-pushup/utils';
import { HistoryCliOptions } from './history.model';

export async function normalizeHashOptions(
  processArgs: HistoryCliOptions & HistoryOptions,
): Promise<HistoryCliOptions & HistoryOptions> {
  const {
    onlySemverTags,
    // overwritten
    maxCount,
    ...opt
  } = processArgs;

  // eslint-disable-next-line functional/no-let, prefer-const
  let { from, to, ...processOptions } = opt;
  // if no semver filter is used resolve hash of tags, as hashes are used to collect history
  if (!onlySemverTags) {
    if (from && isSemver(from)) {
      const { hash } = await getHashFromTag(from);
      from = hash;
    }
    if (to && isSemver(to)) {
      const { hash } = await getHashFromTag(to);
      to = hash;
    }
  }

  return {
    ...processOptions,
    onlySemverTags,
    maxCount: maxCount && maxCount > 0 ? maxCount : undefined,
    from,
    to,
  };
}
