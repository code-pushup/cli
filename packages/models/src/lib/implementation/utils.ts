import type { CategoryConfig } from '../category-config';
import type { PluginConfig } from '../plugin-config';
import type { PluginReport } from '../report';

/**
 * Regular expression to validate a slug for categories, plugins and audits.
 * - audit (e.g. 'max-lines')
 * - category (e.g. 'performance')
 * Also validates ``and ` `
 */
export const slugRegex = /^[a-z\d]+(?:-[a-z\d]+)*$/;

/**
 * Regular expression to validate a filename.
 */
export const filenameRegex = /^(?!.*[ \\/:*?"<>|]).+$/;

/**
 * helper function to validate string arrays
 *
 * @param strings
 */
export function hasDuplicateStrings(strings: string[]): string[] | false {
  const sortedStrings = [...strings].sort();
  const duplStrings = sortedStrings.filter(
    (item, index) => index !== 0 && item === sortedStrings[index - 1],
  );

  return duplStrings.length === 0 ? false : [...new Set(duplStrings)];
}

/**
 * helper function to validate string arrays
 *
 * @param toCheck
 * @param existing
 */
export function hasMissingStrings(
  toCheck: string[],
  existing: string[],
): string[] | false {
  const nonExisting = toCheck.filter(s => !existing.includes(s));
  return nonExisting.length === 0 ? false : nonExisting;
}

/**
 * helper for error items
 */
export function errorItems(
  items: string[] | false,
  transform: (itemArr: string[]) => string = itemArr => itemArr.join(', '),
): string {
  return transform(items || []);
}

export function exists<T>(value: T): value is NonNullable<T> {
  return value != null;
}

/**
 * Get category references that do not point to any audit or group
 * @param categories
 * @param plugins
 * @returns Array of missing references.
 */
export function getMissingRefsForCategories(
  categories: CategoryConfig[],
  plugins: PluginConfig[] | PluginReport[],
) {
  if (categories.length === 0) {
    return false;
  }

  const auditRefsFromCategory = categories.flatMap(({ refs }) =>
    refs
      .filter(({ type }) => type === 'audit')
      .map(({ plugin, slug }) => `${plugin}/${slug}`),
  );
  const auditRefsFromPlugins = plugins.flatMap(({ audits, slug: pluginSlug }) =>
    audits.map(({ slug }) => `${pluginSlug}/${slug}`),
  );
  const missingAuditRefs = hasMissingStrings(
    auditRefsFromCategory,
    auditRefsFromPlugins,
  );

  const groupRefsFromCategory = categories.flatMap(({ refs }) =>
    refs
      .filter(({ type }) => type === 'group')
      .map(({ plugin, slug }) => `${plugin}#${slug} (group)`),
  );
  const groupRefsFromPlugins = plugins.flatMap(({ groups, slug: pluginSlug }) =>
    Array.isArray(groups)
      ? groups.map(({ slug }) => `${pluginSlug}#${slug} (group)`)
      : [],
  );
  const missingGroupRefs = hasMissingStrings(
    groupRefsFromCategory,
    groupRefsFromPlugins,
  );

  const missingRefs = [missingAuditRefs, missingGroupRefs]
    .filter((refs): refs is string[] => Array.isArray(refs) && refs.length > 0)
    .flat();

  return missingRefs.length > 0 ? missingRefs : false;
}

export function missingRefsForCategoriesErrorMsg(
  categories: CategoryConfig[],
  plugins: PluginConfig[] | PluginReport[],
) {
  const missingRefs = getMissingRefsForCategories(categories, plugins);
  return `The following category references need to point to an audit or group: ${errorItems(
    missingRefs,
  )}`;
}
