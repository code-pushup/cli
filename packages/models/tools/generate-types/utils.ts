import { mkdir, stat, writeFile } from 'node:fs/promises';
import { dirname } from 'node:path';
import { ZodTypeAny } from 'zod';
import { createTypeAlias, printNode, zodToTs } from 'zod-to-ts';

export function generateTypesString(
  schemas: Record<string, ZodTypeAny>,
): string {
  return Object.entries(schemas)
    .map(([identifier, schema]) => {
      const typeName = schemaNameToTypeName(identifier);
      const { node } = zodToTs(schema, typeName);
      const typeAlias = createTypeAlias(node, typeName);
      return `export ${printNode(typeAlias)}`;
    })
    .join('\n');
}

export function schemaNameToTypeName(schemaName: string): string {
  const name = schemaName.replace(/Schema/, '');
  return name[0].toUpperCase() + name.slice(1);
}

export function generateMdFile(typesString: string): string {
  return [
    '# Code PushUp config file reference',
    'The `code-pushup.config.(ts|mjs|js)` file should conform to the following type definition:',
    '```ts',
    typesString,
    '```',
  ].join('\n');
}

export async function safeWriteFile(path: string, content: string) {
  const dir = dirname(path) as string;
  const stats = await stat(dir);
  if (!stats.isDirectory()) {
    await mkdir(dir);
  }
  await writeFile(path, content, { encoding: 'utf8' });
}
