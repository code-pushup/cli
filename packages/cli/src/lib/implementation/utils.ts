import { toArray } from '@code-pushup/utils';

// log error and flush stdout so that Yargs doesn't suppress it
// related issue: https://github.com/yargs/yargs/issues/2118
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function logErrorBeforeThrow<T extends (...args: any[]) => any>(
  fn: T,
): T {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (async (...args: any[]) => {
    try {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-argument
      return await fn(...args);
    } catch (error) {
      console.error(error);
      await new Promise(resolve => process.stdout.write('', resolve));
      throw error;
    }
  }) as T;
}

export function coerceArray<T extends string>(param: T | T[] = []): T[] {
  return [
    ...new Set(toArray(param).flatMap((f: T) => f.split(',') as T[]) || []),
  ];
}
