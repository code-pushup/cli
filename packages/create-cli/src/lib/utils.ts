import { readFile, rm, stat, writeFile } from 'node:fs/promises';
import { CODE_PUSHUP_UNICODE_LOGO } from '@code-pushup/utils';
import {
  NX_JSON_CONTENT,
  NX_JSON_FILENAME,
  PROJECT_JSON_CONTENT,
  PROJECT_JSON_FILENAME,
  PROJECT_NAME,
} from './constants.js';

export type SetupResult = {
  filename: string;
  teardown: boolean;
};

export async function setupFile(
  filename: string,
  content = '',
): Promise<SetupResult> {
  const setupResult: SetupResult = {
    filename,
    teardown: false,
  };

  try {
    const stats = await stat(filename);
    if (!stats.isFile()) {
      await writeFile(filename, content);
    }
  } catch (error) {
    if (
      error instanceof Error &&
      error.message.includes('no such file or directory')
    ) {
      await writeFile(filename, content);
      return {
        ...setupResult,
        teardown: true,
      };
    } else {
      console.error(error);
    }
  }

  return setupResult;
}

export function parseNxProcessOutput(output: string) {
  return output.trim().replace('NX', CODE_PUSHUP_UNICODE_LOGO);
}

export async function setupNxContext(): Promise<{
  nxJsonTeardown: boolean;
  projectJsonTeardown: boolean;
  projectName: string;
}> {
  const { teardown: nxJsonTeardown } = await setupFile(
    NX_JSON_FILENAME,
    NX_JSON_CONTENT,
  );
  const { teardown: projectJsonTeardown } = await setupFile(
    PROJECT_JSON_FILENAME,
    PROJECT_JSON_CONTENT,
  );

  const projectJsonContent = await readFile(PROJECT_JSON_FILENAME, 'utf8');
  const { name = PROJECT_NAME } = JSON.parse(projectJsonContent) as {
    name: string;
  };

  return {
    nxJsonTeardown,
    projectJsonTeardown,
    projectName: name,
  };
}

export async function teardownNxContext({
  nxJsonTeardown,
  projectJsonTeardown,
}: {
  nxJsonTeardown: boolean;
  projectJsonTeardown: boolean;
}) {
  const filesToDelete = [
    ...(nxJsonTeardown ? [NX_JSON_FILENAME] : []),
    ...(projectJsonTeardown ? [PROJECT_JSON_FILENAME] : []),
  ];
  await Promise.all(filesToDelete.map(file => rm(file)));
}
