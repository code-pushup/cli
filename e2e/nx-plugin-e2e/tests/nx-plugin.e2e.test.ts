import { Tree } from '@nx/devkit';
import { rm } from 'node:fs/promises';
import { join, relative } from 'node:path';
import { readProjectConfiguration } from 'nx/src/generators/utils/project-configuration';
import { afterEach, expect } from 'vitest';
import { generateCodePushupConfig } from '@code-pushup/nx-plugin';
import {
  generateWorkspaceAndProject,
  materializeTree,
  nxShowProjectJson,
  registerPluginInWorkspace,
} from '@code-pushup/test-nx-utils';
import { removeColorCodes } from '@code-pushup/test-utils';
import { executeProcess, readTextFile } from '@code-pushup/utils';

// @TODO replace with default bin after https://github.com/code-pushup/cli/issues/643
export function relativePathToCwd(testDir: string): string {
  return relative(join(process.cwd(), testDir), process.cwd());
}

describe('nx-plugin', () => {
  let tree: Tree;
  const project = 'my-lib';
  const projectRoot = join('libs', project);
  const baseDir = 'tmp/nx-plugin-e2e/plugin';

  beforeEach(async () => {
    tree = await generateWorkspaceAndProject(project);
  });

  afterEach(async () => {
    await rm(baseDir, { recursive: true, force: true });
  });

  it('should add configuration target dynamically', async () => {
    const cwd = join(baseDir, 'add-configuration-dynamically');
    registerPluginInWorkspace(
      tree,
      join(relativePathToCwd(cwd), 'dist/packages/nx-plugin'),
    );
    await materializeTree(tree, cwd);

    const { code, projectJson } = await nxShowProjectJson(cwd, project);
    expect(code).toBe(0);

    expect(projectJson.targets).toStrictEqual({
      ['code-pushup--configuration']: {
        configurations: {}, // @TODO understand why this appears. should not be here
        executor: 'nx:run-commands',
        options: {
          command: `nx g @code-pushup/nx-plugin:configuration --skipTarget --targetName="code-pushup" --project="${project}"`,
        },
      },
    });

    expect(projectJson).toMatchSnapshot();
  });

  it('should execute dynamic configuration target', async () => {
    const cwd = join(baseDir, 'execute-dynamic-configuration');
    registerPluginInWorkspace(tree, {
      plugin: join(relativePathToCwd(cwd), 'dist/packages/nx-plugin'),
      options: {
        bin: join(relativePathToCwd(cwd), 'dist/packages/nx-plugin'),
      },
    });
    await materializeTree(tree, cwd);

    const { code, stdout } = await executeProcess({
      command: 'npx',
      args: ['nx', 'run', `${project}:code-pushup--configuration`],
      cwd,
    });

    expect(code).toBe(0);

    const cleanStdout = removeColorCodes(stdout);
    expect(cleanStdout).toContain(
      `>  NX   Successfully ran target code-pushup--configuration for project ${project}`,
    );
    await expect(
      readTextFile(join(cwd, projectRoot, 'code-pushup.config.ts')),
    ).resolves.toMatchSnapshot();
  });

  it('should consider plugin option targetName in configuration target', async () => {
    const cwd = join(baseDir, 'configuration-option-target-name');
    registerPluginInWorkspace(tree, {
      plugin: join(relativePathToCwd(cwd), 'dist/packages/nx-plugin'),
      options: {
        targetName: 'cp',
      },
    });
    await materializeTree(tree, cwd);

    const { code, projectJson } = await nxShowProjectJson(cwd, project);

    expect(code).toBe(0);

    expect(projectJson.targets).toStrictEqual({
      ['cp--configuration']: expect.any(Object),
    });
  });

  it('should consider plugin option bin in configuration target', async () => {
    const cwd = join(baseDir, 'configuration-option-bin');
    registerPluginInWorkspace(tree, {
      plugin: join(relativePathToCwd(cwd), 'dist/packages/nx-plugin'),
      options: {
        bin: 'XYZ',
      },
    });
    await materializeTree(tree, cwd);

    const { code, projectJson } = await nxShowProjectJson(cwd, project);

    expect(code).toBe(0);

    expect(projectJson.targets).toStrictEqual({
      ['code-pushup--configuration']: expect.objectContaining({
        options: {
          command: `nx g XYZ:configuration --skipTarget --targetName="code-pushup" --project="${project}"`,
        },
      }),
    });
  });

  it('should NOT add config targets dynamically if the project is configured', async () => {
    const cwd = join(baseDir, 'configuration-already-configured');
    registerPluginInWorkspace(
      tree,
      join(relativePathToCwd(cwd), 'dist/packages/nx-plugin'),
    );
    const { root } = readProjectConfiguration(tree, project);
    generateCodePushupConfig(tree, root);
    await materializeTree(tree, cwd);

    const { code, projectJson } = await nxShowProjectJson(cwd, project);

    expect(code).toBe(0);

    expect(projectJson.targets).toStrictEqual(
      expect.not.objectContaining({
        ['code-pushup--configuration']: expect.any(Object),
      }),
    );
    expect(projectJson).toMatchSnapshot();
  });

  it('should add executor target dynamically if the project is configured', async () => {
    const cwd = join(baseDir, 'add-executor-dynamically');
    registerPluginInWorkspace(
      tree,
      join(relativePathToCwd(cwd), 'dist/packages/nx-plugin'),
    );
    const { root } = readProjectConfiguration(tree, project);
    generateCodePushupConfig(tree, root);
    await materializeTree(tree, cwd);

    const { code, projectJson } = await nxShowProjectJson(cwd, project);
    expect(code).toBe(0);

    expect(projectJson.targets).toStrictEqual({
      ['code-pushup']: {
        configurations: {},
        executor: `@code-pushup/nx-plugin:autorun`,
        options: {},
      },
    });

    expect(projectJson).toMatchSnapshot();
  });

  it('should execute dynamic executor target', async () => {
    const cwd = join(baseDir, 'execute-dynamic-executor');
    const pathRelativeToPackage = relative(join(cwd, 'libs', project), cwd);
    registerPluginInWorkspace(tree, {
      plugin: join(relativePathToCwd(cwd), 'dist/packages/nx-plugin'),
      options: {
        bin: join(relativePathToCwd(cwd), 'dist/packages/nx-plugin'),
      },
    });
    const { root } = readProjectConfiguration(tree, project);
    generateCodePushupConfig(tree, root, {
      fileImports: `import type {CoreConfig} from "${join(
        relativePathToCwd(cwd),
        pathRelativeToPackage,
        'dist/packages/models',
      )}";`,
      plugins: [
        {
          fileImports: `import jsPackagesPlugin from "${join(
            relativePathToCwd(cwd),
            pathRelativeToPackage,
            'dist/packages/plugin-js-packages',
          )}";`,
          // @TODO improve formatObjectToJsString to get rid of the "`" hack
          codeStrings: 'await jsPackagesPlugin({packageManager: `npm`})',
        },
      ],
    });
    await materializeTree(tree, cwd);

    const { stdout, stderr } = await executeProcess({
      command: 'npx',
      args: ['nx', 'run', `${project}:code-pushup -- --dryRun`],
      cwd,
    });

    const cleanStderr = removeColorCodes(stderr);
    // @TODO create test environment for working plugin. This here misses package-lock.json to execute correctly
    expect(cleanStderr).toContain(
      'DryRun execution of: npx @code-pushup/cli autorun',
    );

    const cleanStdout = removeColorCodes(stdout);
    expect(cleanStdout).toContain(
      'NX   Successfully ran target code-pushup for project my-lib',
    );
  });

  it('should consider plugin option bin in executor target', async () => {
    const cwd = join(baseDir, 'configuration-option-bin');
    registerPluginInWorkspace(tree, {
      plugin: join(relativePathToCwd(cwd), 'dist/packages/nx-plugin'),
      options: {
        bin: 'XYZ',
      },
    });
    const { root } = readProjectConfiguration(tree, project);
    generateCodePushupConfig(tree, root);
    await materializeTree(tree, cwd);

    const { code, projectJson } = await nxShowProjectJson(cwd, project);

    expect(code).toBe(0);

    expect(projectJson.targets).toStrictEqual({
      ['code-pushup']: expect.objectContaining({
        executor: 'XYZ:autorun',
      }),
    });
  });

  it('should NOT add targets dynamically if plugin is not registered', async () => {
    const cwd = join(baseDir, 'plugin-not-registered');
    await materializeTree(tree, cwd);

    const { code, projectJson } = await nxShowProjectJson(cwd, project);

    expect(code).toBe(0);

    expect(projectJson.targets).toStrictEqual({});
    expect(projectJson).toMatchSnapshot();
  });
});
