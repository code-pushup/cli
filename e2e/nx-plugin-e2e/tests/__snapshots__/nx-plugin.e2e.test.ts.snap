// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`nx-plugin > should NOT add config targets dynamically if the project is configured 1`] = `
{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "implicitDependencies": [],
  "name": "my-lib",
  "projectType": "library",
  "root": "libs/my-lib",
  "sourceRoot": "libs/my-lib/src",
  "tags": [
    "scope:plugin",
  ],
  "targets": {
    "code-pushup": {
      "configurations": {},
      "executor": "@code-pushup/nx-plugin:autorun",
      "options": {},
    },
  },
}
`;

exports[`nx-plugin > should add configuration target dynamically 1`] = `
{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "implicitDependencies": [],
  "name": "my-lib",
  "projectType": "library",
  "root": "libs/my-lib",
  "sourceRoot": "libs/my-lib/src",
  "tags": [
    "scope:plugin",
  ],
  "targets": {
    "code-pushup--configuration": {
      "configurations": {},
      "executor": "nx:run-commands",
      "options": {
        "command": "nx g @code-pushup/nx-plugin:configuration --skipTarget --targetName="code-pushup" --project="my-lib"",
      },
    },
  },
}
`;

exports[`nx-plugin > should add executor target dynamically if the project is configured 1`] = `
{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "implicitDependencies": [],
  "name": "my-lib",
  "projectType": "library",
  "root": "libs/my-lib",
  "sourceRoot": "libs/my-lib/src",
  "tags": [
    "scope:plugin",
  ],
  "targets": {
    "code-pushup": {
      "configurations": {},
      "executor": "@code-pushup/nx-plugin:autorun",
      "options": {},
    },
  },
}
`;

exports[`nx-plugin > should execute dynamic configuration target 1`] = `
"import type { CoreConfig } from '@code-pushup/models';

// see: https://github.com/code-pushup/cli/blob/main/packages/models/docs/models-reference.md#coreconfig
export default {
  plugins: [],
} satisfies CoreConfig;
"
`;
